notebooks/02-Distributed-Training_comments.ipynb

```diff
@@ -0,0 +1,962 @@
+    "        optimizer.zero_grad()  # reset gradients\n",
+    "        z = model(batch)  # forward pass\n",
+    "        targets = F.one_hot(batch[\"label\"], num_classes=num_classes).float()\n",
+    "        J = loss_fn(z, targets)  # define loss\n",
+    "        J.backward()  # backward pass\n",
+    "        optimizer.step()  # update weights\n",
+    "        loss += (J.detach().item() - loss) / (i + 1)  # cumulative loss\n",
+    "    return loss"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def eval_epoch(ds, batch_size, model, num_classes, loss_fn):\n",
+    "    model.eval()\n",
+    "    loss = 0.0\n",
+    "    y_trues, y_preds = [], []\n",
+    "    ds_generator = ds.iter_torch_batches(batch_size=batch_size, collate_fn=collate_fn)\n",
+    "    with torch.inference_mode():\n",
+    "        for i, batch in enumerate(ds_generator):\n",
+    "            z = model(batch)\n",
+    "            targets = F.one_hot(batch[\"label\"], num_classes=num_classes).float()  # one-hot (for loss_fn)\n",
+    "            J = loss_fn(z, targets).item()\n",
+    "            loss += (J - loss) / (i + 1)\n",
+    "            y_trues.extend(batch[\"label\"].cpu().numpy())\n",
+    "            y_preds.extend(torch.argmax(z, dim=1).cpu().numpy())\n",
+    "    return loss, np.vstack(y_trues), np.vstack(y_preds)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def train_loop_per_worker(config):\n",
+    "    # Hyperparameters\n",
+    "    model_registry = config[\"model_registry\"]\n",
+    "    experiment_name = config[\"experiment_name\"]\n",
+    "    embedding_dim = config[\"embedding_dim\"]\n",
+    "    hidden_dim = config[\"hidden_dim\"]\n",
+    "    dropout_p = config[\"dropout_p\"]\n",
+    "    lr = config[\"lr\"]\n",
+    "    lr_factor = config[\"lr_factor\"]\n",
+    "    lr_patience = config[\"lr_patience\"]\n",
+    "    num_epochs = config[\"num_epochs\"]\n",
+    "    batch_size = config[\"batch_size\"]\n",
+    "    num_classes = config[\"num_classes\"]\n",
+    "\n",
+    "    # Experiment tracking\n",
+    "    if ray.train.get_context().get_world_rank() == 0:\n",
+    "        mlflow.set_tracking_uri(f\"file:{model_registry}\")\n",
+    "        mlflow.set_experiment(experiment_name)\n",
+    "        mlflow.start_run()\n",
+    "        mlflow.log_params(config)\n",
+    "\n",
+    "    # Datasets\n",
+    "    train_ds = ray.train.get_dataset_shard(\"train\")\n",
+    "    val_ds = ray.train.get_dataset_shard(\"val\")\n",
+    "\n",
+    "    # Model\n",
+    "    model = ClassificationModel(\n",
+    "        embedding_dim=embedding_dim, \n",
+    "        hidden_dim=hidden_dim, \n",
+    "        dropout_p=dropout_p, \n",
+    "        num_classes=num_classes,\n",
+    "    )\n",
+    "    model = ray.train.torch.prepare_model(model)\n",
```

Comment:
Let's actually just wrap the model directly with `DistributedDataParallel` instead of using this ray train utility. We are planning to de-emphasize this API.